site/server/app/com/fortysevendeg/exercises/persistence/domain/UserProgressQueries.scala

```diff
@@ -5,47 +5,55 @@
 
 package com.fortysevendeg.exercises.persistence.domain
 
+import shared.UserProgress
+import doobie.imports._
+
 object SaveUserProgress {
 
-  sealed trait ExerciseType
+  sealed abstract class ExerciseType extends Product with Serializable
 
   case object Koans extends ExerciseType
 
   case object Other extends ExerciseType
 
   case class Request(
-    userId:       Long,
-    libraryName:  String,
-    sectionName:  String,
-    method:       String,
-    version:      Int,
-    exerciseType: ExerciseType   = Other,
-    args:         Option[String],
-    succeeded:    Boolean
-  )
+      userId:       Long,
+      libraryName:  String,
+      sectionName:  String,
+      method:       String,
+      version:      Int,
+      exerciseType: ExerciseType   = Other,
+      args:         Option[String],
+      succeeded:    Boolean
+  ) {
+
+    def asUserProgress(id: Long): UserProgress =
+      UserProgress(id, userId, libraryName, sectionName, method, version, exerciseType.toString, args, succeeded)
+  }
 
 }
 
 object UserProgressQueries {
 
-  val allFields = List("userid", "libraryname", "sectionname", "method", "version", "exercisetype", "args", "succeeded")
+  val allFields = List("id", "userid", "libraryname", "sectionname", "method", "version", "exercisetype", "args", "succeeded")
```

Comment:
We could have had this auto derived with shapeless records given the case class structure.